msal/application.py

```diff
@@ -1333,9 +1349,10 @@ def _acquire_token_silent_from_cache_and_possibly_refresh_it(
             force_refresh=False,  # type: Optional[boolean]
             claims_challenge=None,
             correlation_id=None,
+            auth_scheme=None,
             **kwargs):
         access_token_from_cache = None
-        if not (force_refresh or claims_challenge):  # Bypass AT when desired or using claims
+        if not (force_refresh or claims_challenge or auth_scheme):  # Then attempt AT cache
```

Comment:
I am not sure if this is the case, but if POP auth scheme is specified, we should also not attempt to use RT in the cache. This will break migration scenario: 

1. App uses webview and points to a cache file. App gets bearer tokens for a user: AT, RT, IdT
2. App switches to WAM, while still pointing at the same cache file, so as to not lose state, i.e. GetAccounts() correctly returns 1 account.
3. App calls AcquireTokenSilent(user) + POP scheme

MSAL now: 

- bypasses the AT token cache 
- looks in the cache for an RT
- finds the RT and ?? I'm not sure what happens here, but we should not use the RT or fail. We should just go to WAM.

Maybe this is a test you want to write? 

PS: we have this behavior now in MSAL.NET and it is one of the GA blockers.

Reply:
> if POP auth scheme is specified, we should also not attempt to use RT in the cache. This will break migration scenario:
> ...
> Maybe this is a test you want to write?
>
> PS: we have this behavior now in MSAL.NET and it is one of the GA blockers.

Nice call-out. MSAL Python implementation does not have such an issue, because the subsequent code path will use broker and will not fall back.

I will see how to add a test case here.

P.S.: The other day, our downstream partner asked whether MSAL .Net is already GA now. Is it?

Reply:
UPDATE: [Added a new check](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/511/files#diff-ed3b8af48cd21dc1ac735b839d89b172a6f45569c324af70a2bdb4f60e5fbcddR130-R132) in the existing test helpers. Closing this conversation thread.